File: clients/go/offline.go
package zama

import (
	"context"
	"errors"
	"math/big"
	"time"

	"github.com/ethereum/go-ethereum/common"
	pb "github.com/zama-ai/sdk/clients/go/v3/internal/gen/zama/sdk/v1beta1"
	"google.golang.org/grpc"
)

// TransactionRequest is sealed; only the request types in this package implement it.
type TransactionRequest interface {
	prepareWire() (*pb.PrepareTransactionRequest, error)
}

type PrepareOptions struct {
	Nonce    *uint64
	GasLimit *big.Int
	Fees     *PrepareFees
}
type PrepareFees struct {
	MaxFeePerGas         *big.Int
	MaxPriorityFeePerGas *big.Int
}

// TransactionKind mirrors the wire enum; its values are the wire numbers.
type TransactionKind int32

const (
	TransactionConfidentialTransfer     = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_CONFIDENTIAL_TRANSFER)
	TransactionConfidentialTransferFrom = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_CONFIDENTIAL_TRANSFER_FROM)
	TransactionSetOperator              = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_SET_OPERATOR)
	TransactionUnwrap                   = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_UNWRAP)
	TransactionUnwrapAll                = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_UNWRAP_ALL)
	TransactionFinalizeUnwrap           = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_FINALIZE_UNWRAP)
	TransactionApproveUnderlying        = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_APPROVE_UNDERLYING)
	TransactionWrap                     = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_WRAP)
	TransactionTransferAndCall          = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_TRANSFER_AND_CALL)
	TransactionDelegateDecryption       = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_DELEGATE_DECRYPTION)
	TransactionRevokeDelegation         = TransactionKind(pb.TransactionKind_TRANSACTION_KIND_REVOKE_DELEGATION)
)

func (k TransactionKind) String() string { return pb.TransactionKind(k).String() }

func transactionKind(wire pb.TransactionKind) (TransactionKind, error) {
	if wire == pb.TransactionKind_TRANSACTION_KIND_UNSPECIFIED {
		return 0, errors.New("prepared transaction has no kind")
	}
	if _, known := pb.TransactionKind_name[int32(wire)]; !known {
		return 0, errors.New("prepared transaction has an unknown kind")
	}
	return TransactionKind(wire), nil
}

type PreparedTransaction struct {
	Kind       TransactionKind
	From       common.Address
	UnsignedTx []byte
}

// decimal rejects nil so a missing value never reaches the wire as an empty string.
func decimal(name string, value *big.Int) (string, error) {
	if value == nil {
		return "", errors.New(name + " is required")
	}
	return value.String(), nil
}
func prepareOptionsWire(options *PrepareOptions) (*pb.PrepareOptions, error) {
	if options == nil {
		return nil, nil
	}
	result := &pb.PrepareOptions{Nonce: options.Nonce}
	if options.GasLimit != nil {
		value := options.GasLimit.String()
		result.GasLimit = &value
	}
	if options.Fees != nil {
		maxFee, err := decimal("max fee per gas", options.Fees.MaxFeePerGas)
		if err != nil {
			return nil, err
		}
		priority, err := decimal("max priority fee per gas", options.Fees.MaxPriorityFeePerGas)
		if err != nil {
			return nil, err
		}
		result.Fees = &pb.PrepareFees{MaxFeePerGas: maxFee, MaxPriorityFeePerGas: priority}
	}
	return result, nil
}

// PrepareTransaction returns an unsigned transaction for the caller to sign and broadcast.
func (s *SDKContext) PrepareTransaction(ctx context.Context, request TransactionRequest, options *PrepareOptions) (PreparedTransaction, error) {
	if request == nil {
		return PreparedTransaction{}, errors.New("transaction request is required")
	}
	wire, err := request.prepareWire()
	if err != nil {
		return PreparedTransaction{}, err
	}
	if wire.Options, err = prepareOptionsWire(options); err != nil {
		return PreparedTransaction{}, err
	}
	response, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.PrepareTransactionResponse, error) {
		wire.Operation = op
		return s.client.rpc.PrepareTransaction(ctx, wire, trailer)
	})
	if err != nil {
		return PreparedTransaction{}, err
	}
	from, err := addressFromWire(response.From, "prepared transaction has invalid sender length")
	if err != nil {
		return PreparedTransaction{}, err
	}
	kind, err := transactionKind(response.Kind)
	if err != nil {
		return PreparedTransaction{}, err
	}
	return PreparedTransaction{Kind: kind, From: from, UnsignedTx: response.UnsignedTx}, nil
}

type ConfidentialTransferRequest struct {
	From   common.Address
	Token  common.Address
	To     common.Address
	Amount *big.Int
}

func (r ConfidentialTransferRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	amount, err := decimal("amount", r.Amount)
	if err != nil {
		return nil, err
	}
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_ConfidentialTransfer{ConfidentialTransfer: &pb.ConfidentialTransfer{Token: r.Token.Bytes(), To: r.To.Bytes(), Amount: amount}}}, nil
}

type ConfidentialTransferFromRequest struct {
	From   common.Address
	Token  common.Address
	Owner  common.Address
	To     common.Address
	Amount *big.Int
}

func (r ConfidentialTransferFromRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	amount, err := decimal("amount", r.Amount)
	if err != nil {
		return nil, err
	}
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_ConfidentialTransferFrom{ConfidentialTransferFrom: &pb.ConfidentialTransferFrom{Token: r.Token.Bytes(), Owner: r.Owner.Bytes(), To: r.To.Bytes(), Amount: amount}}}, nil
}

type SetOperatorRequest struct {
	From     common.Address
	Token    common.Address
	Operator common.Address
	// Until is a required Unix timestamp in seconds; a positive past value revokes.
	Until *uint64
}

func (r SetOperatorRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	if r.Until == nil {
		return nil, errors.New("until is required")
	}
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_SetOperator{SetOperator: &pb.SetOperator{Token: r.Token.Bytes(), Operator: r.Operator.Bytes(), Until: r.Until}}}, nil
}

type UnwrapRequest struct {
	From   common.Address
	Token  common.Address
	To     common.Address
	Amount *big.Int
}

func (r UnwrapRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	amount, err := decimal("amount", r.Amount)
	if err != nil {
		return nil, err
	}
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_Unwrap{Unwrap: &pb.Unwrap{Token: r.Token.Bytes(), To: r.To.Bytes(), Amount: amount}}}, nil
}

type UnwrapAllRequest struct {
	From  common.Address
	Token common.Address
	To    common.Address
}

func (r UnwrapAllRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_UnwrapAll{UnwrapAll: &pb.UnwrapAll{Token: r.Token.Bytes(), To: r.To.Bytes()}}}, nil
}

type FinalizeUnwrapRequest struct {
	From                    common.Address
	Wrapper                 common.Address
	UnwrapRequestIDOrAmount []byte
}

func (r FinalizeUnwrapRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_FinalizeUnwrap{FinalizeUnwrap: &pb.FinalizeUnwrap{Wrapper: r.Wrapper.Bytes(), UnwrapRequestIdOrAmount: r.UnwrapRequestIDOrAmount}}}, nil
}

type ApproveUnderlyingRequest struct {
	From       common.Address
	Underlying common.Address
	Spender    common.Address
	Amount     *big.Int
}

func (r ApproveUnderlyingRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	amount, err := decimal("amount", r.Amount)
	if err != nil {
		return nil, err
	}
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_ApproveUnderlying{ApproveUnderlying: &pb.ApproveUnderlying{Underlying: r.Underlying.Bytes(), Spender: r.Spender.Bytes(), Amount: amount}}}, nil
}

type WrapRequest struct {
	From    common.Address
	Wrapper common.Address
	To      common.Address
	Amount  *big.Int
}

func (r WrapRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	amount, err := decimal("amount", r.Amount)
	if err != nil {
		return nil, err
	}
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_Wrap{Wrap: &pb.Wrap{Wrapper: r.Wrapper.Bytes(), To: r.To.Bytes(), Amount: amount}}}, nil
}

type TransferAndCallRequest struct {
	From       common.Address
	Underlying common.Address
	Wrapper    common.Address
	Amount     *big.Int
	// Nil omits recipient data; an empty non-nil slice sends explicit empty bytes.
	RecipientData []byte
}

func (r TransferAndCallRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	amount, err := decimal("amount", r.Amount)
	if err != nil {
		return nil, err
	}
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_TransferAndCall{TransferAndCall: &pb.TransferAndCall{Underlying: r.Underlying.Bytes(), Wrapper: r.Wrapper.Bytes(), Amount: amount, RecipientData: r.RecipientData}}}, nil
}

type DelegateDecryptionRequest struct {
	From            common.Address
	ContractAddress common.Address
	DelegateAddress common.Address
	ExpirationDate  *time.Time
}

func (r DelegateDecryptionRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	delegation, err := delegateDecryptionWire(r.ContractAddress, r.DelegateAddress, r.ExpirationDate)
	if err != nil {
		return nil, err
	}
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_DelegateDecryption{DelegateDecryption: delegation}}, nil
}

type RevokeDelegationRequest struct {
	From            common.Address
	ContractAddress common.Address
	DelegateAddress common.Address
}

func (r RevokeDelegationRequest) prepareWire() (*pb.PrepareTransactionRequest, error) {
	delegation := revokeDelegationWire(r.ContractAddress, r.DelegateAddress)
	return &pb.PrepareTransactionRequest{From: r.From.Bytes(), Transaction: &pb.PrepareTransactionRequest_RevokeDelegation{RevokeDelegation: delegation}}, nil
}

File: clients/rust/src/offline.rs
use crate::{Address, BigInt, Offline, generated};
use anyhow::{Context, Result, bail};

#[derive(Clone, Debug)]
pub struct PrepareTransaction {
    pub from: Address,
    pub transaction: Transaction,
}

#[derive(Clone, Debug)]
pub enum Transaction {
    ConfidentialTransfer {
        token: Address,
        to: Address,
        amount: BigInt,
    },
    ConfidentialTransferFrom {
        token: Address,
        owner: Address,
        to: Address,
        amount: BigInt,
    },
    SetOperator {
        token: Address,
        operator: Address,
        /// Unix timestamp in whole seconds; a positive past value revokes the operator.
        until: u64,
    },
    Unwrap {
        token: Address,
        to: Address,
        amount: BigInt,
    },
    UnwrapAll {
        token: Address,
        to: Address,
    },
    FinalizeUnwrap {
        wrapper: Address,
        unwrap_request_id_or_amount: Vec<u8>,
    },
    ApproveUnderlying {
        underlying: Address,
        spender: Address,
        amount: BigInt,
    },
    Wrap {
        wrapper: Address,
        to: Address,
        amount: BigInt,
    },
    TransferAndCall {
        underlying: Address,
        wrapper: Address,
        amount: BigInt,
        recipient_data: Option<Vec<u8>>,
    },
    DelegateDecryption {
        contract_address: Address,
        delegate_address: Address,
        /// Unix time in whole milliseconds; omission grants permanent delegation.
        expiration_date_ms: Option<u64>,
    },
    RevokeDelegation {
        contract_address: Address,
        delegate_address: Address,
    },
}

#[derive(Clone, Debug, Default)]
pub struct PrepareOptions {
    pub nonce: Option<u64>,
    pub gas_limit: Option<BigInt>,
    pub fees: Option<PrepareFees>,
}

#[derive(Clone, Debug)]
pub struct PrepareFees {
    pub max_fee_per_gas: BigInt,
    pub max_priority_fee_per_gas: BigInt,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum TransactionKind {
    ConfidentialTransfer,
    ConfidentialTransferFrom,
    SetOperator,
    Unwrap,
    UnwrapAll,
    FinalizeUnwrap,
    ApproveUnderlying,
    Wrap,
    TransferAndCall,
    DelegateDecryption,
    RevokeDelegation,
}

impl TryFrom<i32> for TransactionKind {
    type Error = anyhow::Error;

    fn try_from(value: i32) -> Result<Self> {
        let Ok(kind) = generated::TransactionKind::try_from(value) else {
            bail!("unknown prepared transaction kind");
        };
        Ok(match kind {
            generated::TransactionKind::Unspecified => {
                bail!("unspecified prepared transaction kind")
            }
            generated::TransactionKind::ConfidentialTransfer => Self::ConfidentialTransfer,
            generated::TransactionKind::ConfidentialTransferFrom => Self::ConfidentialTransferFrom,
            generated::TransactionKind::SetOperator => Self::SetOperator,
            generated::TransactionKind::Unwrap => Self::Unwrap,
            generated::TransactionKind::UnwrapAll => Self::UnwrapAll,
            generated::TransactionKind::FinalizeUnwrap => Self::FinalizeUnwrap,
            generated::TransactionKind::ApproveUnderlying => Self::ApproveUnderlying,
            generated::TransactionKind::Wrap => Self::Wrap,
            generated::TransactionKind::TransferAndCall => Self::TransferAndCall,
            generated::TransactionKind::DelegateDecryption => Self::DelegateDecryption,
            generated::TransactionKind::RevokeDelegation => Self::RevokeDelegation,
        })
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PreparedTransaction {
    pub kind: TransactionKind,
    pub from: Address,
    pub unsigned_tx: Vec<u8>,
}

impl Offline {
    /// Prepares through the SDK; the caller owns signing and broadcasting the returned bytes.
    pub async fn prepare(
        &self,
        request: PrepareTransaction,
        options: Option<PrepareOptions>,
    ) -> Result<PreparedTransaction> {
        let result = rpc!(
            &self.0,
            prepare_transaction,
            PrepareTransactionRequest {
                from: request.from.to_vec(),
                transaction: Some(request.transaction.into()),
                options: options.map(Into::into),
            }
        )
        .await?;
        Ok(PreparedTransaction {
            kind: TransactionKind::try_from(result.kind)?,
            from: Address::try_from(result.from.as_slice())
                .context("invalid prepared sender address")?,
            unsigned_tx: result.unsigned_tx,
        })
    }
}

impl From<PrepareOptions> for generated::PrepareOptions {
    fn from(value: PrepareOptions) -> Self {
        Self {
            nonce: value.nonce,
            gas_limit: value.gas_limit.map(|n| n.to_string()),
            fees: value.fees.map(|fees| generated::PrepareFees {
                max_fee_per_gas: fees.max_fee_per_gas.to_string(),
                max_priority_fee_per_gas: fees.max_priority_fee_per_gas.to_string(),
            }),
        }
    }
}

impl From<Transaction> for generated::prepare_transaction_request::Transaction {
    fn from(value: Transaction) -> Self {
        match value {
            Transaction::ConfidentialTransfer { token, to, amount } => {
                Self::ConfidentialTransfer(generated::ConfidentialTransfer {
                    token: token.to_vec(),
                    to: to.to_vec(),
                    amount: amount.to_string(),
                })
            }
            Transaction::ConfidentialTransferFrom {
                token,
                owner,
                to,
                amount,
            } => Self::ConfidentialTransferFrom(generated::ConfidentialTransferFrom {
                token: token.to_vec(),
                owner: owner.to_vec(),
                to: to.to_vec(),
                amount: amount.to_string(),
            }),
            Transaction::SetOperator {
                token,
                operator,
                until,
            } => Self::SetOperator(generated::SetOperator {
                token: token.to_vec(),
                operator: operator.to_vec(),
                until: Some(until),
            }),
            Transaction::Unwrap { token, to, amount } => Self::Unwrap(generated::Unwrap {
                token: token.to_vec(),
                to: to.to_vec(),
                amount: amount.to_string(),
            }),
            Transaction::UnwrapAll { token, to } => Self::UnwrapAll(generated::UnwrapAll {
                token: token.to_vec(),
                to: to.to_vec(),
            }),
            Transaction::FinalizeUnwrap {
                wrapper,
                unwrap_request_id_or_amount,
            } => Self::FinalizeUnwrap(generated::FinalizeUnwrap {
                wrapper: wrapper.to_vec(),
                unwrap_request_id_or_amount,
            }),
            Transaction::ApproveUnderlying {
                underlying,
                spender,
                amount,
            } => Self::ApproveUnderlying(generated::ApproveUnderlying {
                underlying: underlying.to_vec(),
                spender: spender.to_vec(),
                amount: amount.to_string(),
            }),
            Transaction::Wrap {
                wrapper,
                to,
                amount,
            } => Self::Wrap(generated::Wrap {
                wrapper: wrapper.to_vec(),
                to: to.to_vec(),
                amount: amount.to_string(),
            }),
            Transaction::TransferAndCall {
                underlying,
                wrapper,
                amount,
                recipient_data,
            } => Self::TransferAndCall(generated::TransferAndCall {
                underlying: underlying.to_vec(),
                wrapper: wrapper.to_vec(),
                amount: amount.to_string(),
                recipient_data,
            }),
            Transaction::DelegateDecryption {
                contract_address,
                delegate_address,
                expiration_date_ms,
            } => Self::DelegateDecryption(generated::DelegateDecryption {
                contract_address: contract_address.to_vec(),
                delegate_address: delegate_address.to_vec(),
                expiration_date_ms,
            }),
            Transaction::RevokeDelegation {
                contract_address,
                delegate_address,
            } => Self::RevokeDelegation(generated::RevokeDelegation {
                contract_address: contract_address.to_vec(),
                delegate_address: delegate_address.to_vec(),
            }),
        }
    }
}
